### ISSUE TITLE
Can only generate changesets once?

### ISSUE THREAD

**Issue by adbutterfield** [iamchathu/changeset-conventional-commits#34]
I'm looking at integrating this into the release flow for a new monorepo I'm working on.

I generated the changesets once locally, but then deleted them all (still working on the bugs of the workflow). But then when I run `pnpm dlx changeset-conventional-commits` again, I don't get anything generated...

Is this expected behavior? (Or maybe I'm doing something stupid?)

**Comment by adbutterfield** [comment]
Maybe I figured it out. Seems like it checks current branch against base branch (upstream main) and looks for commits. I have commit that haven't been published in any version, but the are in both branches. 

I think that's it...

**Comment by mblackrittr** [comment]
> I generated the changesets once locally, but then deleted them all (still working on the bugs of the workflow). But then when I run `pnpm dlx changeset-conventional-commits` again, I don't get anything generated...

Hmmmm, if they get deleted, they'll generate again, because the (de)duplication-check `difference()` (and, yeah, I also agree, it should be named more expressive) will detect nothing.

> Is this expected behavior? (Or maybe I'm doing something stupid?)

It's surely and 100% more likely the lacking verbosity, but we're on it! 😻

**Comment by adbutterfield** [comment]
I know the first time I got the message "No git tags found, using repo's first commit for automated change detection. Note: this may take a while.", but now I don't get that message. There's also still no tags.

**Comment by adbutterfield** [comment]
`git describe --tags --abbrev=0` for me prints `fatal: No names found, cannot describe anything.`.

**Comment by mblackrittr** [comment]
> Maybe I figured it out. Seems like it checks current branch against base branch (upstream main) and looks for commits. I have commit that haven't been published in any version, but the are in both branches.

I'll keep this in mind for the next MR-/PR-updates on [feat: add cli helper, logging and flags --dry, --git-fetch and --verbose #29](https://github.com/iamchathu/changeset-conventional-commits/pull/29) - to give some verbosity/output also on these matters.

> I think that's it...

I'm not sure, but thanks for the reminder - will run a few scenarios when checking for more verbosity on these matters, and yeah, setting up some test-cases for these would be nice, will see.

But there're ideas to change the fixed behavior, to gain more control over it and make it adjustable to various scenarios, like already having a tag, but it's not release related, so still wanting to generate `changesets` since before this tag, etc., because right now it determines the "newness" of commits on the latest tag - in case I'm not myself already fully lost on this complexity, haha! 😹💦

In case you make more findings, maybe you can elaborate some details by lining out the scenario in a few steps - maybe with references to the involved/affecting code?

**Comment by adbutterfield** [comment]
Idk actually. I just checked out `main` from the feature branch that wasn't generating the changesets the second time, and hit `pnpm dlx changeset-conventional-commits`, and it made the changesets...

**Comment by mblackrittr** [comment]
Nice to hear, so it was just `something` - because beside the actual `changesets` and doing a `git fetch origin ${branch}` (`baseBranch` from `.changeset/config.json`), [CCC] doesn't modify or store anything itself (so far) - so by deleting the `changesets` and without any external changes, it gives the same results again.

But there're surely a handful circumstances where it can "blow" (or, well, just do "nothing") on unexpected reasons - private packages, no workspaces, diverging branches after rebase or force, interfering non-release-tags, ... - but rescue's on its way and I'm adding a bunch of more "knobs" and verbosity to the mentioned MR/PR which should take place next days and finally end the `voodoo guessing`! 😻

As a little background on my involvement with `CCC`: It's an integral part of my [DevOps-Monorepo] [^1] "meta-package", so I should be around with this for a while, whereby my "humble" aspiration is to make it either become part of [Changesets] itself or at least the de-facto and go-to solution to integrate `Conventional Commits/Changelogs` with `Changesets` - making `Changesets` a superset of everything `Conventional`, **#OhHellboyYeah!** haha 🥳🎉

[CCC]: https://github.com/iamchathu/changeset-conventional-commits/
[Changesets]: https://github.com/changesets/changesets/
[DevOps-Monorepo]: https://github.com/mblackrittr/mbr-demo/

[^1]: Yeah, still finding the right name, so `MBR-Demo` (started as just a small demo), will surely change when it's finally leaving `0.0.0` - probably something like `Monorepo-DevOps` should fit it.

**Comment by adbutterfield** [comment]
I wonder if there's still a bug (or just something not ideal?) in there somewhere. I can consistently re-generate on `main`, but I don't get them generated on feature branches I cut off of `main`. (Maybe because it considers local `main` the base branch instead of `origin/main`?)

Anyways, safe to close for now I think.

**Comment by mblackrittr** [comment]
***Oh!*** 🙀

Ah yeah, right, well, that's indeed expected and intended behavior:
> If this is running on a branch that was created from the main branch, we want to get commits all **since the branch was created**.

https://github.com/iamchathu/changeset-conventional-commits/blob/50bf163d24b0edfc3c9b8b1ab284ef61fbd07522/src/utils/index.ts#L141-L143

Beside that's surely to be considered a sane concept - so it doesn't pull in `commits` and generates `changesets` not intended to - it's missing documentation and hints (verbosity) while running, so this is also noted for update!

Thanks man, for nicely completing the mentioned "handful circumstances", because I'd just listed four! 😹💦

**Comment by iamchathu** [comment]
Yeah, I will slowly add more documentation. This code currently ignore the root package changes. So it not currently works with changeset and non mono-repo setups. We need to update code to work in both cases.

**Comment by mblackrittr** [comment]
No worries, it's already all in - plenty of options, logging, docs, Changesets' config recognition and regular/non-monorepo support **and** root (`/`) support for monorepos (which we'll still've to implement into the real Changesets), plus surely some more - just finishing and splitting up this more or less complete rewrite into some sane commits! 😹💦
